### DISCUSSION TITLE
Returning unscheduled Tasks

### DISCUSSION THREAD

**Discussion by fredrikekre** [PartitionedArrays/PartitionedArrays.jl#61] [General]
I noticed that some `async_*` routines return unscheduled task. I am just curious about this choice (`@task` over `@async`) since I have not seen this pattern elsewhere. I can understand that it gives the user a choice of when to schedule these tasks, but it seems like you would in 99% of cases schedule them right away. Is there a practical reason when you would want to delay the scheduling?

Also, in e.g. the `test_fem_sa.jl`]() example, this task
https://github.com/fverdugo/PartitionedArrays.jl/blob/d5224f7c188436da103b3883e13a0c5f6eca64b0/test/test_fem_sa.jl#L74-L76
is not scheduled until later
https://github.com/fverdugo/PartitionedArrays.jl/blob/d5224f7c188436da103b3883e13a0c5f6eca64b0/test/test_fem_sa.jl#L103-L104

I was confused by the comment there since the task is not actually started yet. Would it be better to use this pattern in this example (and elsewhere):
```julia
t = map_parts(schedule, async_assemble!(I, J, V, rows))

# do the other stuff here

 map_parts(wait, t)
```

**Comment by fverdugo** [comment]
> I noticed that some async_* ..

Yes, there is a reason.This is one of the parts that is not yet well resolved as you pointed out. So, any new idas will be welcomed!

The main reason is that the mpi process will block as soon as you schedule the task. This is a quite tricky problem. See the discussion here

 https://discourse.julialang.org/t/how-to-combine-mpi-non-blocking-isend-irecv-and-julia-tasks/52524

for details.

MPI.jl has an open PR that could help to improve the API at the PartitionedArrays.jl side: https://github.com/JuliaParallel/MPI.jl/pull/452

**Comment by fredrikekre** [comment]
Thanks for the pointers.

So right now there is not really a point in using `async_assemble!` instead of `assemble!`, right? E.g. in the example I linked above, there it nothing on the lines between that needs to happen before scheduling the task?

**Comment by fverdugo** [comment]
> So right now there is not really a point in using async_assemble! instead of assemble!, right?

There is a point. When you call async_assemble! the MPI communication begins right away and you can overlap computations until you call map_parts(wait∘schedule,t). Note that this schedules and waits for a Julia task that wrapps the MPI wait. But the MPI communication itself started before.

**Comment by fredrikekre** [comment]
Ah I see now that not everything is wrapped in the returned `@task`, e.g. this:
https://github.com/fverdugo/PartitionedArrays.jl/blob/d5224f7c188436da103b3883e13a0c5f6eca64b0/src/Interfaces.jl#L2479
is happening before `async_assemble!` returns, and this:
https://github.com/fverdugo/PartitionedArrays.jl/blob/d5224f7c188436da103b3883e13a0c5f6eca64b0/src/Interfaces.jl#L2509
is not scheduled until later. So `async_assemble!` is partly asynchronous.

**Comment by fverdugo** [comment]
This discussion is not relevant anymore. Since v0.3.0, the library always returns scheduled tasks.
